hooks/useScrollPosition.tsx

```diff
@@ -0,0 +1,120 @@
+import { use, useEffect, useRef, useState } from "react";
+
+type ElementItems = {
+  maxScrollHeight: number;
+  maxScrollWidth: number; // will be the max scroll width
+  height: number;
+  width: number;
+  horizontalScroll: number;
+  verticalScroll: number;
+} | null;
+
+// give the html element of the scroll container
+export const useScrollPosition = (prop: string) => {
```

Comment:
I feel like this approach is a little much....

The scroll indicator will only be present on the horizontal gallery screen and only used for horizontal scrolling, so no need to worry about vertical scroll. 

The component you made is fine,but  why not just have it take a prop that is a ref of the horizontalGallery its going to represent, and have it do these calculations itself?

Reply:
Def is a bit much if it'll only be used once 😆 . My real goal here was to learn about the react api. I do think a hook to handle the users scroll position could have its uses though. As I mentioned in the discord, this is not even useful in its current state 😢

Reply:
```
import React, { useState, useRef } from 'react';

const ModalGallery = () => {
  const [scrollPosition, setScrollPosition] = useState(0);
  const horizontalGalleryRef = useRef(null);

  const handleScroll = () => {
    if (horizontalGalleryRef.current) {
      setScrollPosition(horizontalGalleryRef.current.scrollLeft);  // Tracks horizontal scroll
    }
  };

  return (
    <div>
      // Pass handleScroll and ref to HorizontalGallery
      <HorizontalGallery handleScroll={handleScroll} ref={horizontalGalleryRef} />
      
      // Pass scrollposition as prop
      <ScrollIndicator scrollPosition={scrollPosition} />
    </div>
  );
};
```

Then horizontalGallery component forwards the ref to the scrollable div and calls the handleScroll function when the user scrolls.

Then scrollIndicator receives the scrollPosition as a prop and can render or act based on that value.

Reply:
Does this work? I'm curious how the the `handleScroll` is actually handled on the `HorizontalGallery`, when it runs etc. 

If `ref` here is anything like Vue then I would assume any changes, including scrolling, will cause the ref value to update so this is def a more efficient way to get the scroll position of an element instead of using a document selector. 

However I am not convinced this snippet will actually work without some changes. I might be mistaken, but I believe the `ref` will need to go on the scroll container regardless, otherwise I am not sure how it will know what is being ran. 

Is the `handleScroll` on the gallery an event listener? When does this run, or is this also provided by react?

Reply:
You probably don't even need a ref at all. All we are trying to do is observe the scrollPosition

Maybe just have  HorizontalGallery take handleScroll  prop and thats it. 
`const HorizontalGallery = ({ onScroll }) =>  { ... }`

Then you can forward handleScroll func to the scrollable div inside HorizontalGallery
` <div onScroll={handleScroll} style={{ overflowY: 'hidden', overflowX: 'scroll'}} id="scroll-container">`.

Now the parent (ModalGallery) gets the scroll position directly from the scroll event in the HorizontalGallery component, without needing a ref.

Reply:
Forward the ref and func to root div of HorizontalGallery

```
const HorizontalGallery = React.forwardRef(({ handleScroll }, ref) => {
  return (
    <div  ref={ref} 
      onScroll={handleScroll} 
     style={{ overflowY: 'hidden', overflowX: 'scroll'}} id="scroll-container"
    >
    ...
    </div>
```

Because you are forwarding the ref to it using React.forwardRef, When the component is rendered, the DOM element (our scrollable div) is assigned to ref.current, and in your parent component, that ref is horizontalGalleryRef. We are also forwarding handle scroll to the same div, so any time it scrolls we are calling handleScroll in the parent. There, we will have all the information we need to write a function that returns our scroll percentage. 

Then, we just pass percentage value as a prop to the scrollIndicator

Reply:
Okay damn! Thats nice. Makes sense that is provided to us too. Thanks for calling out.